keps/sig-api-machinery/3352-aggregated-discovery/README.md

```diff
@@ -0,0 +1,758 @@
+
+The most well known Kubernetes client that uses the discovery mechanism is `kubectl`, and more specifically the `CachedDiscoveryClient` in `client-go`. To mitigate some of this latency, kubectl has implemented a (10 minute timer) 6 hours(!) timer during which the discovery API is not refreshed. The drawback of this approach is that the freshness of the cache is doubtful and the entire discovery API needs to be refreshed after 6 hours, even if it hasn’t expired.
+
+This not only impacts kubectl, but all clients of kubernetes. We can do better.
+
+### Goals
+
+- Fix the discovery storm issue currently present in kubectl
+- Aggregate the discovery documents for all Kubernetes types
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+- Since the current discovery separated by group-version is already GA, removal of the endpoint will be difficult. This KEP will solely focus on introducing the new aggregated endpoint and will not cover deprecation.
+
+## Proposal
+
+We are proposing a new endpoint `/discovery` as an aggregated endpoint for all discovery documents. Discovery documents can currently be found under `apis/<group>/<version>` and `/api/v1` for the legacy group version. This discovery endpoint will support publishing an ETag so clients who already have the latest version of the aggregated discovery can avoid redownloading the document.
+
+We will add a new controller responsible for aggregating the discovery documents when a resource on the cluster changes. There will be no conflicts when aggregating since each discovery document is self-contained.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+We will expose a endpoint `/discovery` that will support JSON, and
```

Comment:
how will we version this API so that we don't burn the name when we (probably) need a v3?

Reply:
How about something like `/discovery/v1` similar to how we version the OpenAPI?

Reply:
> How about something like /discovery/v1 similar to how we version the OpenAPI?

That would work for me.